### ISSUE TITLE
Using AWS Cognito as OAuth2 authenticator and querying the Groups field

### ISSUE THREAD

**Issue by FleischerT** [trinodb/trino#15027]
Hi,

i have set up Aws Cognito as my Oauth2 authenticator and that works as expected. But now I'm trying to set up catalog level access control for which I need the Cognito groups, so I set the groups field:

http-server.authentication.oauth2.groups-field=cognito:groups

but that has no effect. After looking at Trino's OAuth2 implementation, I think the problem is that Cognito embeds the group information in the id_token instead of the access_token. 
Does anyone have a workaround for this issue or are there plans to support AWS Cognito as an authenticator in the future?

**Comment by Praveen2112** [comment]
cc: @lukasz-walkiewicz , @s2lomon

**Comment by lkm** [comment]
I don't believe this is uncommon, to the best of my knowledge `access_token` in the token response is not expected to be a JWT, but `id_token` MUST be, see https://openid.net/specs/openid-connect-core-1_0.html#TokenResponse

If we wanted to avoid BC break this should at the very least be configuable but problably default to use `id_token` as the token to validate and read keys from.

**Comment by FleischerT** [comment]
Any updates on this?

**Comment by rolanddb** [comment]
I believe this is the case for Azure AD as well; the `groups` claim is available in the `id_token`. 

In any case, setting `http-server.authentication.oauth2.groups-field=groups` does not lead to an error, but I don't see the user being assigned to a group, either..

**Comment by FleischerT** [comment]
> I believe this is the case for Azure AD as well; the `groups` claim is available in the `id_token`.
> 
> In any case, setting `http-server.authentication.oauth2.groups-field=groups` does not lead to an error, but I don't see the user being assigned to a group, either..

This assumption is correct. For example, see the implementation in Airflow for Azure AD:
https://github.com/apache/airflow/blob/fdac67b3a5350ab4af79fd98612592511ca5f3fc/airflow/www/fab_security/manager.py#L691

**Comment by rolanddb** [comment]
I can see from the AAD response that we get a regular access token and not an ID token. Is there any way to control `response_type`? Because Trino is asking for `code` rather than `id_token`.

**Comment by dancook-doxo** [comment]
Having set up Cognito-Trino integration, would it be possible to expose a new Trino catalog showing Cognito user pools? In other words Cognito becomes a provider of catalogs/data to Trino?

**Comment by FleischerT** [comment]
> Having set up Cognito-Trino integration, would it be possible to expose a new Trino catalog showing Cognito user pools? In other words Cognito becomes a provider of catalogs/data to Trino?

I'm not quite sure I understand what you want. But I think you want to read the cognito user pools with trino. 
I don't think that is possible at the moment.

**Comment by dancook-doxo** [comment]
Yes, you've correctly understood what I'm trying to do. I will continue to
investigate but thanks for the reply.

**Comment by rolanddb** [comment]
> .

I can see how a search for "cognito + groups" would lead someone to this issue, but I think that's a completely different issue. This issue is about the groups claim in the token.

**Comment by FleischerT** [comment]
Any updates on this?
